README.md
[![GitHub Workflow Status](https://img.shields.io/github/workflow/status/mandelbrot-project/memo/CI%20Build)](https://github.com/mandelbrot-project/memo/actions)
![GitHub](https://img.shields.io/github/license/mandelbrot-project/memo?color=blue)
[![PyPI](https://img.shields.io/pypi/v/memo_ms?color=blue)](https://pypi.org/project/memo-ms/)

# memo
**M**s2 bas**E**d sa**M**ple vect**O**rization (**MEMO**) package

## Description

MEMO is a method allowing a Retention Time (RT) agnostic alignment of metabolomics samples using the fragmentation spectra (MS2) of their consituents.

The occurence of MS2 Peaks and Neutral Losses in each sample is counted and used to generate an *MS2 fingerprint* of the sample. These fingerprints can in a second stage be aligned to compare different samples. Once obtained, different filtering (remove peaks/losses from blanks for example) and visualization techniques (MDS/PCoA, TMAP, Heatmap, ...) can be used. 

MEMO suits particularly well to compare chemodiverse samples, ie with a poor features overlap, or to compare samples with a strong RT shift, acquired using different LC methods or even different mass spectrometers technology (Maxiis qToF vs Q-Exactive).

### Small preview of the results on the Qemistree Evaluation Dataset:

The Tutorial uses the [Qemistree Evaluation Dataset](https://www.nature.com/articles/s41589-020-00677-3). This dataset contains samples analyzed on a Q-Exactive using 2 different LC methods simulating a really strong batch effect. Samples originates from 2 fecal samples, 1 tomato sample & 1 plasma sample (+ binary and quaternary mixtures of these 4 samples).

#### Samples colored according to their content
![plot](./pcoa_tuto_contains.png)

#### Samples colored according to their LC method
![plot](./pcoa_tuto_method.png)

#### Samples clustered using HCA
![plot](./hca.png)

#### Samples clustered using HCA and Heatmap visualization
![plot](./heatmap.png)



MEMO is mainly built on [matchms](https://github.com/matchms/matchms) and [spec2vec](https://github.com/iomega/spec2vec) packages for handling the MS2 spectra and convert them into documents. Huge thanks to them for the amazing work done with these packages!

### To install it:

First make sure to have [anaconda](https://www.anaconda.com/products/individual) installed.

#### A) Using pip install

A.1. Create a new conda environment to avoid clashes:
```
conda create --name memo python=3.8
conda activate memo
```
A.2. Install with pip:
```
pip install numpy
pip install memo-ms
```

If you have an error, try insstalling scikit-bio from conda-forge before installing the package with pip:
```
conda install -c conda-forge scikit-bio
pip install memo-ms
```

You can clone the repository to get the demo spectra and quant table files!

#### B) Clone and install locally
B.1. First clone the repository using git clone in command line:
```
git clone <ssh_key or https>
```
B.2. Create a new conda environment to avoid clashes:
```
conda create --name memo python
conda activate memo
```

B.3. Install the package locally using pip:
```
pip install .
```

#### Enjoy!

Once installed, try it using the tutorial jupyter notebook.

Please add feedback to the issues section. Thanks !

## Documentation for developers

#### Installation
Create an environment with
```
git clone https://github.com/mandelbrot-project/memo.git
cd memo
conda create --name memo-dev python=3.8
conda activate memo-dev
```
Then install dependencies and memo:
```
python -m pip install --upgrade pip
pip install numpy
pip install --editable .[dev]
# pip install -e .'[dev]' (on mac)
```

#### Run tests
Memo tests can be run by:
```
pytest
``` 
And the code linter with
```
prospector
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Unreleased

### Added

- Minimal script to perform memo on a mzmine feature table (CSV) with related metadata (CSV) and spectral file (MGF) (tutorial/minimal_memo.py)

## [0.1.4] - 2021-12-16

### Added

- Filtering option to filter peaks/losses present in more/less thant N percent of samples [#25](https://github.com/mandelbrot-project/memo/issues/25)
- Possibility to build memo_from_unaligned() from files matching pattern even if in subfolders

### Changed

- actions: add readme web-link checker
- actions: developed test_user_workflow.py

## [0.1.3] - 2021-12-16

### Added

- Documentation

## [0.1.2] - 2021-12-15

### Added

- test for memo_from_unaligned
- test for memo_from_aligned
- test for merged_memo()
- possibility to import "memo ready" feature table

### Changed

- Changed classes.py to avoid erasing when filtering
- Changed .removesuffix to allow support of python <3.9

## [0.1.1] - 2021-12-10

### Changed

- Changed required version of python from >3.8 to >3.7 for TMAP compatibility

## [0.1.0] - 2021-12-06

### Added

- New `MemoMatrix` class, replacing the former `MemoContainer` [#14](https://github.com/matchms/matchms/pull/14)
- First set of unit tests [#12](https://github.com/matchms/matchms/pull/12)
- Continuous integration for testing [#4](https://github.com/matchms/matchms/pull/4)

### Changed

- Code linting [#16](https://github.com/matchms/matchms/pull/16)
- Code linting [#14](https://github.com/matchms/matchms/pull/14)

## [0.0.4] - 2021-11-01

### Added

- This is the initial version of memo from https://github.com/mandelbrot-project/memo


[Unreleased]: https://github.com/mandelbrot-project/memo/compare/0.1.0...HEAD
[0.1.3]: https://github.com/mandelbrot-project/memo/compare/0.1.2...0.1.3
[0.1.2]: https://github.com/mandelbrot-project/memo/compare/0.1.1...0.1.2
[0.1.1]: https://github.com/mandelbrot-project/memo/compare/0.1.0...0.1.1
[0.1.0]: https://github.com/mandelbrot-project/memo/compare/0.0.4...0.1.0
[0.0.4]: https://github.com/mandelbrot-project/memo/releases/tag/0.0.4


tutorial/tutorial_memo.py
# %% [markdown]
# # MEMO Tutorial on the Qemistree Evaluation Dataset

# %% [markdown]
# In this tutorial, we will use the Qemistree published dataset (https://doi.org/10.1038/s41589-020-00677-3) to apply the MS2 BasEd SaMple VectOrization (MEMO) method
# 
# The dataset is constitued of 2 fecal samples, 1 tomato sample and 1 plasma sample, plus different binary/quaternary mixtures of these four samples. Samples were profiled in UHPLC-MS/MS (Q-Exactive spectrometer) using 2 different LC-methods. Each sample was acquired in triplicates using each LC-method (see Qemistree paper for details).

# %% [markdown]
# ## First we import the needed packages
# Be sure to have first installed memo using 'pip install memo' within the memo environment. 
# Also make sure to launch this notebook using the memo conda environement.

# %%
import pandas as pd
import numpy as np
import memo_ms as memo
import plotly.express as px

# %%
# this cell is tagged 'parameters'
fake_parameter = 'foo'

# %% [markdown]
# With this step we import metadata:

# %%
pwd()

# %%

def conditions(df_meta):
    if ((df_meta['Proportion_Fecal_1']>0) & (df_meta['Proportion_Fecal_2']==0)& (df_meta['Proportion_Tomato']==0) & (df_meta['Proportion_NIST_1950_SRM']==0)):
        return 'Fecal_1'
    if ((df_meta['Proportion_Fecal_1']==0) & (df_meta['Proportion_Fecal_2']>0)& (df_meta['Proportion_Tomato']==0) & (df_meta['Proportion_NIST_1950_SRM']==0)):
        return 'Fecal_2'
    if ((df_meta['Proportion_Fecal_1']==0) & (df_meta['Proportion_Fecal_2']==0)& (df_meta['Proportion_Tomato']>0) & (df_meta['Proportion_NIST_1950_SRM']==0)):
        return 'Tomato'
    if ((df_meta['Proportion_Fecal_1']==0) & (df_meta['Proportion_Fecal_2']==0)& (df_meta['Proportion_Tomato']==0) & (df_meta['Proportion_NIST_1950_SRM']>0)):
        return 'Plasma'
    if ((df_meta['Proportion_Fecal_1']>0) & (df_meta['Proportion_Fecal_2']>0)& (df_meta['Proportion_Tomato']==0) & (df_meta['Proportion_NIST_1950_SRM']==0)):
        return 'Fecal_1 + Fecal_2'
    if ((df_meta['Proportion_Fecal_1']>0) & (df_meta['Proportion_Fecal_2']==0)& (df_meta['Proportion_Tomato']>0) & (df_meta['Proportion_NIST_1950_SRM']==0)):
        return 'Fecal_1 + Tomato'
    if ((df_meta['Proportion_Fecal_1']>0) & (df_meta['Proportion_Fecal_2']==0)& (df_meta['Proportion_Tomato']==0) & (df_meta['Proportion_NIST_1950_SRM']>0)):
        return 'Fecal_1 + Plasma'
    if ((df_meta['Proportion_Fecal_1']==0) & (df_meta['Proportion_Fecal_2']>0)& (df_meta['Proportion_Tomato']>0) & (df_meta['Proportion_NIST_1950_SRM']==0)):
        return 'Fecal_2 + Tomato'
    if ((df_meta['Proportion_Fecal_1']==0) & (df_meta['Proportion_Fecal_2']>0)& (df_meta['Proportion_Tomato']==0) & (df_meta['Proportion_NIST_1950_SRM']>0)):
        return 'Fecal_2 + Plasma'
    if ((df_meta['Proportion_Fecal_1']==0) & (df_meta['Proportion_Fecal_2']==0)& (df_meta['Proportion_Tomato']>0) & (df_meta['Proportion_NIST_1950_SRM']>0)):
        return 'Tomato + Plasma'
    if ((df_meta['Proportion_Fecal_1']>0) & (df_meta['Proportion_Fecal_2']>0)& (df_meta['Proportion_Tomato']>0) & (df_meta['Proportion_NIST_1950_SRM']>0)):
        return 'Fecal_1 + Fecal_2 + Tomato + Plasma' 
    else:
        return 'What is it? :)'


df_meta = pd.read_csv("../data/1901_gradient_benchmarking_dataset_v4_sample_metadata.txt", sep='\t')
df_meta['Samplename'] = df_meta['Samplename'].str[:-6]
df_meta['Samplename'] = df_meta['Samplename'].str.replace('BLANK_', 'BLANK')
df_meta = df_meta[['Filename', 'Experiment', 'Samplename', 'Triplicate_number', 'Proportion_Fecal_1', 'Proportion_Fecal_2', 'Proportion_Tomato', 'Proportion_NIST_1950_SRM']]
df_meta['contains'] = df_meta.apply(conditions, axis=1)
df_meta['instrument'] = np.where(df_meta['Samplename'].str.contains('qTOF'), 'qTOF', 'QE')
df_meta['blank_qc'] = np.where(df_meta['Samplename'].str.contains('blank|qcmix', case = False), 'yes', 'no')
df_meta


# %% [markdown]
# ## Import feature_quant table
# To compute the MEMO matrix of the dataset, we need the table reporting presence/absence of each metabolite in each sample. This information is in the quant table and we create a memo.FeatureTable dataclass object to load it.

# %%
feat_table_qe = memo.FeatureTable(path="../data/quantification_table-00000.csv", software='mzmine')
feat_table_qe.feature_table

# %% [markdown]
# ## Import spectra
# Since MEMO relies on the occurence of MS2 fragments/losses in samples to compare them, we obviously need to importthe features' fragmentation spectra. Losses are computed and spectra translated into documents. Store in memo.SpectraDocuments dataclass object.

# %%
spectra_qe = memo.SpectraDocuments(path="../data/qemistree_specs_ms.mgf", min_relative_intensity = 0.01,
            max_relative_intensity = 1, min_peaks_required=5, losses_from = 10, losses_to = 200, n_decimals = 2)
spectra_qe.document

# %% [markdown]
# ## Generation of MEMO matrix
# 
# Using the generated documents and the quant table, we can now obtain the MEMO matrix. The MEMO matrix is stored in the MemoContainer object, along with the feature table and the documents

# %%
memo_qe = memo.MemoMatrix()
memo_qe.memo_from_aligned_samples(feat_table_qe, spectra_qe)
memo_qe.memo_matrix

# %%
memo_qe = memo_qe.filter(use_samples_pattern=True, samples_pattern='blank')
feat_table_qe = feat_table_qe.filter(use_samples_pattern=True, samples_pattern='blank')
memo_qe.memo_matrix

# %% [markdown]
# 

# %% [markdown]
# ## Plotting
# 
# We can now use the MEMO matrix to generate the PCoA of our samples

# %%

memo.visualization.plot_pcoa_2d(
    matrix= memo_qe.memo_matrix,
    df_metadata=df_meta,
    metric= 'braycurtis',
    filename_col = 'Filename',
    group_col='contains',
    norm = False,
    scaling= False,
    pc_to_plot = [1,2]
)

# %%
memo.visualization.plot_pcoa_3d(
    matrix= memo_qe.memo_matrix,
    df_metadata=df_meta,
    metric= 'braycurtis',
    filename_col = 'Filename',
    group_col='contains',
    norm = False,
    scaling= False,
    pc_to_plot = [1,2,3]
)

# %%
df_meta_sub = df_meta[df_meta['Triplicate_number'] == 1]

memo.visualization.plot_heatmap(
    matrix= memo_qe.memo_matrix,
    df_metadata=df_meta_sub,
    filename_col = 'Filename',
    group_col = 'contains',
    plotly_discrete_cm = px.colors.qualitative.Plotly,
    linkage_method='ward',
    linkage_metric = 'euclidean',
    heatmap_metric = 'braycurtis',
    norm = False,
    scaling= False
)

# %%
memo.visualization.plot_hca(
    matrix= memo_qe.memo_matrix,
    df_metadata=df_meta_sub,
    filename_col = 'Filename',
    group_col = 'contains',
    linkage_method='ward',
    linkage_metric = 'euclidean',
    norm = False,
    scaling= False
)

# %% [markdown]
# ## Merge MEMO matrix from different MzMine projects

# %% [markdown]
# First, we load as before the spectra and feature matrix to generate the memo matrix of the different projects to merge.
# 
# In this case, we will once again use the Qemistree dataset and compare for the same samples data aquired on a QToF with some on a Q-Exactive (Orbitrap).

# %%
# Generate MEMO matrix for QE data

feat_table_qe = memo.FeatureTable(path="../data/qe_qtof_coanalysis/qe_quant_nogapF.csv", software='mzmine')
spectra_qe = memo.SpectraDocuments(path="../data/qe_qtof_coanalysis/qe_spectra_nogapF.mgf", min_relative_intensity = 0.01,
            max_relative_intensity = 1, min_peaks_required=5, losses_from = 10, losses_to = 200, n_decimals = 2)
memo_qe = memo.MemoMatrix()
memo_qe.memo_from_aligned_samples(feat_table_qe, spectra_qe)

# %%
# Generate MEMO matrix for QToF data

feat_table_qtof = memo.FeatureTable(path="../data/qe_qtof_coanalysis/qtof_quant_nogapF.csv", software='mzmine')
spectra_qtof = memo.SpectraDocuments(path="../data/qe_qtof_coanalysis/qtof_spectra_nogapF.mgf", min_relative_intensity = 0.01,
            max_relative_intensity = 1, min_peaks_required=5, losses_from = 10, losses_to = 200, n_decimals = 2)
memo_qtof = memo.MemoMatrix()
memo_qtof.memo_from_aligned_samples(feat_table_qtof, spectra_qtof)

# %%
memo_qe = memo_qe.filter(use_samples_pattern=True,samples_pattern='blank')
memo_qtof = memo_qtof.filter(use_samples_pattern=True,samples_pattern='blank')

# %% [markdown]
# Now, let's merge our 2 MEMO matrix. There is one paramater, drop_not_in_common, to decide wether to keep only shared peaks/losses or all of them.

# %%
memo_merged = memo_qe.merge_memo(memo_qtof, drop_not_in_common=True)

# %% [markdown]
# PCoA plot of the first 2 components show that the main parameter to discriminate our samples is the Instrument

# %%

memo.visualization.plot_pcoa_2d(
    matrix= memo_merged.memo_matrix,
    df_metadata=df_meta,
    metric= 'braycurtis',
    filename_col = 'Filename',
    group_col='contains',
    norm = True,
    scaling= False,
    pc_to_plot = [1,2]
)

# %% [markdown]
# But when looking at component 2 and 3, we see that it is possible to cluster chemically related samples using this approach.

# %%

memo.visualization.plot_pcoa_2d(
    matrix= memo_merged.memo_matrix,
    df_metadata=df_meta,
    metric= 'braycurtis',
    filename_col = 'Filename',
    group_col='contains',
    norm = True,
    scaling= False,
    pc_to_plot = [2,3]
)

# %%



